### PR TITLE
DevStudio : ware board + ware migration

### PR DESCRIPTION
**Pull request by Arthoni**
## Status
- The two main features are stabilized
- A lot to do in refinements and code cleaning (cf blocking tags remaining in code)

## UI

Board with error:
image

Board with warning:
image

Board with no error:
image

Migration setup:
image

Migration progress (when succeeding):
image

Migration progress (when failing):
image

### CODE REVIEW

File: src/openfluid/ui/waresdev/GroupStatusMessagesWidget.cpp [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+  Other Usage means a use of OpenFLUID that is inconsistent with the GPL
+  license, and requires a written agreement between You and INRA.
+  Licensees for Other Usage of OpenFLUID may use this file in accordance
+  with the terms contained in the written agreement between You and INRA.
+  
+*/
+
+
+/**
+  @file GroupStatusMessagesWidget.cpp
```
Comment (jctophefabre) [inline src/openfluid/ui/waresdev/GroupStatusMessagesWidget.cpp:34]: Adjust file name for better consistency with other files

File: src/openfluid/ui/waresdev/SignalMigrationListener.hpp [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+  Other Usage means a use of OpenFLUID that is inconsistent with the GPL
+  license, and requires a written agreement between You and INRA.
+  Licensees for Other Usage of OpenFLUID may use this file in accordance
+  with the terms contained in the written agreement between You and INRA.
+  
+*/
+
+
+/**
+  @file SignalMigrationListener.hpp
```
Comment (jctophefabre) [inline src/openfluid/ui/waresdev/SignalMigrationListener.hpp:34]: Adjust file name, something like `UIMigrationListener.hpp` ?

File: src/openfluid/ui/waresdev/SignalMigrationListener.hpp [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+#include <iomanip>
+
+#include <openfluid/waresdev/WareSrcMigratorListener.hpp>
+#include <openfluid/tools/Console.hpp>
+
+
+namespace openfluid { namespace ui { namespace waresdev {
+
+
+class SignalMigrationListener : public QObject, public openfluid::waresdev::WareSrcMigratorListener
```
Comment (jctophefabre) [inline src/openfluid/ui/waresdev/SignalMigrationListener.hpp:54]: Is this class implementation consistent with similar classes in DevStudio?

File: src/openfluid/ui/waresdev/StatusButtonMessageWidget.cpp [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+  Other Usage means a use of OpenFLUID that is inconsistent with the GPL
+  license, and requires a written agreement between You and INRA.
+  Licensees for Other Usage of OpenFLUID may use this file in accordance
+  with the terms contained in the written agreement between You and INRA.
+  
+*/
+
+
+/**
+  @file StatusButtonMessageWidget.cpp
```
Comment (jctophefabre) [inline src/openfluid/ui/waresdev/StatusButtonMessageWidget.cpp:34]: Adjust file name for better consistency with other files

File: src/openfluid/ui/waresdev/StatusButtonMessageWidget.cpp [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+    ui->StatusIconLabel->setWarningStatus();
+  }
+  else if (Level == ReportingData::ReportingStatus::ERROR)
+  {
+    ui->StatusIconLabel->setErrorStatus();
+  }
+  else if (Level == ReportingData::ReportingStatus::DISABLED)
+  {
+      ui->StatusIconLabel->setDisabledStatus();
+  }
```
Comment (jctophefabre) [inline src/openfluid/ui/waresdev/StatusButtonMessageWidget.cpp:160]: Is a default value necessary?

File: src/openfluid/ui/waresdev/WareDashboardStatusWidget.cpp [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+  Other Usage means a use of OpenFLUID that is inconsistent with the GPL
+  license, and requires a written agreement between You and INRA.
+  Licensees for Other Usage of OpenFLUID may use this file in accordance
+  with the terms contained in the written agreement between You and INRA.
+  
+*/
+
+
+/**
+  @file WareDashboardStatusWidget.cpp
```
Comment (jctophefabre) [inline src/openfluid/ui/waresdev/WareDashboardStatusWidget.cpp:34]: Adjust file name for better consistency with other files

File: src/openfluid/ui/waresdev/WareDashboardStatusWidget.hpp [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+
+/**
+  @file WareDashboardStatusWidget.hpp
+
+  @author Jean-Christophe FABRE <jean-christophe.fabre@inrae.fr>
+  @author Armel THÖNI <armel.thoni@inrae.fr>
+ */
+
+
+#ifndef __OPENFLUID_UIWARESDEV_DSDASHBOARDSTATUSWIDGET_HPP__
```
Comment (jctophefabre) [inline src/openfluid/ui/waresdev/WareDashboardStatusWidget.hpp:41]: Wrong header guard

File: src/openfluid/ui/waresdev/WareSrcWidgetCollection.cpp [hunk 8]
```diff
@@ -176,42 +185,140 @@ void WareSrcWidgetCollection::onCloseWareTabRequested(int Index)
...
+    connect(MWorker, SIGNAL(finished(bool, const QString&)), &ProgressDialog,
+            SLOT(finish(bool, const QString&)));
+    connect(MWorker, SIGNAL(info(const QString&)), &ProgressDialog,
+            SLOT(writeInfo(const QString&)));
+    connect(MWorker, SIGNAL(error(const QString&)), &ProgressDialog,
+            SLOT(writeError(const QString&)));
+    connect(MWorker, SIGNAL(progressed(const int)), &ProgressDialog,
+            SLOT(progress(const int)));
+
+    Thread->start();
```
Comment (jctophefabre) [inline src/openfluid/ui/waresdev/WareSrcWidgetCollection.cpp:309]: Thread joining needed?

File: src/openfluid/ui/waresdev/WareSrcWidgetCollection.cpp [hunk 9]
```diff
@@ -176,42 +185,140 @@ void WareSrcWidgetCollection::onCloseWareTabRequested(int Index)
...
+  MigrationWorker* MWorker = new MigrationWorker(WarePath, MigrationSetup.isVerbose(), MigrationSetup.isNewBranchChecked());
+
+  try
+  {
+
+    QThread* Thread = new QThread();
+
+    MWorker->moveToThread(Thread);
+
+    connect(Thread, SIGNAL(started()), MWorker, SLOT(run()));
```
Comment (jctophefabre) [inline src/openfluid/ui/waresdev/WareSrcWidgetCollection.cpp:295]: Disconnection of signal/slots needed once finished?

File: src/openfluid/ui/waresdev/WareSrcWidgetCollection.hpp [hunk 10]
```diff
@@ -51,7 +51,11 @@
 #include <openfluid/ware/SimulatorSignature.hpp>
 #include <openfluid/waresdev/WareBuildOptions.hpp>
 #include <openfluid/waresdev/WareSrcContainer.hpp>
+#include <openfluid/ui/waresdev/GitUIProxy.hpp>
 #include <openfluid/ui/waresdev/FindReplaceDialog.hpp>
+#include <openfluid/waresdev/WareSrcMigrator.hpp>
+
+#include "SignalMigrationListener.hpp"
```
Comment (jctophefabre) [inline src/openfluid/ui/waresdev/WareSrcWidgetCollection.hpp:58]: Wrong include form, use `#include <openfluid/ui/waresdev/SignalMigrationListener.hpp>` instead

File: src/openfluid/ui/waresdev/SignalMigrationListener.hpp (hunk 3 above)
Comment (Arthoni) [reply src/openfluid/ui/waresdev/SignalMigrationListener.hpp:54]: This class could be compared to `RunSimulationListener`, current implementation shows significant differences, see if could be more consistent

File: src/openfluid/ui/waresdev/WareSrcWidgetCollection.cpp (hunk 9 above)
Comment (Arthoni) [reply src/openfluid/ui/waresdev/WareSrcWidgetCollection.cpp:295]: Not done in any thread connect operation, I don't know the impact on performances

File: src/openfluid/waresdev/WareSrcMigratorListener.hpp [hunk 11]
```diff
@@ -50,6 +50,25 @@ namespace openfluid { namespace waresdev {
 
 class WareSrcMigratorListener : public openfluid::base::Listener
 {
+  protected:
+
+    static constexpr std::string_view DISPLAY_TEXT_PRELIMINARY_CHECKS = "Preliminary checks";
```
Comment (jctophefabre) [inline src/openfluid/waresdev/WareSrcMigratorListener.hpp:55]: This class should be agnostic of its subclasses implementations. Strings should not be stored here.

File: src/openfluid/waresdev/WareSrcMigratorListener.hpp [hunk 12]
```diff
@@ -50,6 +50,25 @@ namespace openfluid { namespace waresdev {
 
 class WareSrcMigratorListener : public openfluid::base::Listener
```
Comment (jctophefabre) [inline src/openfluid/waresdev/WareSrcMigratorListener.hpp:51]: This class should be agnostic of its subclasses implementations. Strings should not be stored here.

File: src/openfluid/waresdev/WareSrcMigratorListener.hpp (hunk 11 above)
Comment (jctophefabre) [inline src/openfluid/waresdev/WareSrcMigratorListener.hpp:55]: Why using a std::string_view here?

File: src/openfluid/ui/waresdev/WareSrcWidget.hpp [hunk 13]
```diff
@@ -160,6 +168,8 @@ class OPENFLUID_API WareSrcWidget: public QWidget
     void buildLaunched(openfluid::ware::WareType Type, const QString& ID);
 
     void buildFinished(openfluid::ware::WareType Type, const QString& ID);
+    
+    void migrationRequestedOnWare(const QString& WarePath);
```
Comment (jctophefabre) [inline src/openfluid/ui/waresdev/WareSrcWidget.hpp:172]: Is this the right place for this action?

File: src/openfluid/ui/waresdev/WareStatusDashboardWidget.cpp [hunk 14]
```diff
@@ -0,0 +1,231 @@
...
+// =====================================================================
+// =====================================================================
+
+
+void WareStatusDashboardWidget::setLevel(const ReportingData::ReportingStatus Level)
+{
+  if (Level == ReportingData::ReportingStatus::OK)
+  {
+    ui->TitleLabel->setText(tr("No issue"));
+    m_TitleColor = "#55B332";
```
Comment (jctophefabre) [inline src/openfluid/ui/waresdev/WareStatusDashboardWidget.cpp:199]: Colors values already defined elsewhere (in Builder?).
Should be variablelized?

File: src/openfluid/ui/waresdev/WareStatusItemWidget.hpp [hunk 15]
```diff
@@ -0,0 +1,108 @@
...
+
+class WareStatusItemWidget : public QWidget
+{
+  Q_OBJECT;
+
+  private:
+
+    typedef openfluid::waresdev::WareSrcChecker::ReportingData ReportingData;
+
+    static std::map<std::pair<std::string, ReportingData::ReportingStatus>,
```
Comment (jctophefabre) [inline src/openfluid/ui/waresdev/WareStatusItemWidget.hpp:67]: Coding style for static members?

File: src/openfluid/ui/waresdev/WareStatusItemWidget.cpp [hunk 16]
```diff
@@ -0,0 +1,202 @@
...
+             QString>  WareStatusItemWidget::s_LabelForReportItem;
+std::map<std::pair<std::string, ReportingData::ReportingStatus>, 
+             std::pair<QString, QString>> WareStatusItemWidget::s_ActionForReportItem;
+
+
+// =====================================================================
+// =====================================================================
+
+
+void WareStatusItemWidget::populateReportItemLabels()
```
Comment (jctophefabre) [inline src/openfluid/ui/waresdev/WareStatusItemWidget.cpp:62]: To be completed

File: src/openfluid/ui/waresdev/WareMigrationListener.cpp [hunk 17]
```diff
@@ -0,0 +1,324 @@
...
+namespace openfluid { namespace ui { namespace waresdev {
+
+
+void WareMigrationListener::flushMsg(const std::string& channel)
+{
+  if (channel == "info")
+  {
+    info(m_CurrentMsg);
+  }
+  else if (channel == "warnin")
```
Comment (jctophefabre) [inline src/openfluid/ui/waresdev/WareMigrationListener.cpp:56]: There is a missing "g" at the end of the "warning" word. Is that normal?

File: src/openfluid/ui/waresdev/WareMigrationListener.cpp [hunk 18]
```diff
@@ -0,0 +1,324 @@
...
+#include <iostream>
+#include <iomanip>
+
+#include <openfluid/ui/waresdev/WareMigrationListener.hpp>
+
+
+namespace openfluid { namespace ui { namespace waresdev {
+
+
+void WareMigrationListener::flushMsg(const std::string& channel)
```
Comment (jctophefabre) [inline src/openfluid/ui/waresdev/WareMigrationListener.cpp:50]: Argument name should start with a capital

File: src/openfluid/ui/waresdev/WareSrcWidget.hpp [hunk 19]
```diff
@@ -96,6 +99,12 @@ class OPENFLUID_API WareSrcWidget: public QWidget
 
     openfluid::ui::waresdev::WareSrcUIContainer m_Container;
 
+    std::map<std::string, int> m_TabIndexByName;
+
+    openfluid::ui::waresdev::WareStatusDashboardWidget* mp_Board;
```
Comment (jctophefabre) [inline src/openfluid/ui/waresdev/WareSrcWidget.hpp:104]: Why this are needed? Should not be included in the `ui->` object?

**Pull request merged**